datastore support for endpoint, GetObject, DeleteObject and swarm store API upgrade

mavenugo opened the pull request:
There are a few quirks when it comes to data persistence using the store. on-disk data persistence will be dealt with in subsequent PRs (it may make use of the store or not... depending on the capabilities provided by the store implementation).
Other than that, one can use the dnet tool to test the datastore across multiple-host with their preferred backend (as supported by swarm store).

mrjana commented on controller.go:
```diff
@@ -238,87 +209,52 @@ func (c *controller) NewNetwork(networkType, name string, options ...NetworkOpti
-	// Create the network
-	if err := d.CreateNetwork(network.id, network.generic); err != nil {
+
+	if err := c.addNetworkToStore(network); err != nil {
```
What should we do if `addNetworkToStore` fails? Should we roll back by letting the driver to delete network and removing the entry from local cache? Or just let the network be still available as a local entity?

mavenugo replied:
Good point. we currently dont have the distributed error handler that we discussed and hence addressing this use-case alone is not sufficient to handle the global consistent view between the local cache, programmed state and the distributed store states. the distributed error handler will be addressed as a next step in this series of distributed store commits.

mrjana commented on datastore/datastore.go:
```diff
@@ -37,23 +50,35 @@ type KV interface {
 }
 
+//UnKey provides convenient method to unpack the key to compliment the Key function
+func UnKey(key string) ([]string, error) {
```
I don't think the name `Unkey` is intuitive to let us know what operation is being done. A more appropriate name would be `ParseKey`

mrjana commented on datastore/datastore.go:
```diff
@@ -37,23 +50,35 @@ type KV interface {
+	if len(chain) <= len(rootChain) || !reflect.DeepEqual(chain[0:len(rootChain)], rootChain) {
+		return nil, types.BadRequestErrorf("Invalid Key : %s", key)
+	}
+	return chain[len(rootChain):], nil
```
When you split a key like this `/docker/libnetwork/a/b/` you would get the slice `chain`'s first element to be empty string and the last element to be empty string. So when you return `chain[len(rootChain):]` wouldn't it return the path element `libnetwork` and then the rest of the key and then lastly an empty string. This makes it distinctly asymmetric with the `Key` method above. Ideally when I do `Unkey(Key(<list_of_keys>))` we should be able to get the same list back. Not some thing else right?

mavenugo replied:
You are right. this needs `strings.Trim` for `/`
But fortunately, this didn't impact us as the keys were already trimmed off these. But we should not be relying on that. Will add the `strings.Trim` in the next update.

mrjana commented on datastore/datastore.go:
```diff
@@ -37,23 +50,35 @@ type KV interface {
+func UnKey(key string) ([]string, error) {
+	chain := strings.Split(key, "/")
+	if len(chain) <= len(rootChain) || !reflect.DeepEqual(chain[0:len(rootChain)], rootChain) {
+		return nil, types.BadRequestErrorf("Invalid Key : %s", key)
```
error string should start in lower case

mrjana commented on driverapi/driverapi.go:
```diff
@@ -14,6 +14,9 @@ type Driver interface {
 	Config(options map[string]interface{}) error
 
+	// Capabilities of the driver
+	Capabilities() Capability
```
IMO this should not be a separate api that can be invoked by libnetwork to get the capability every time. It should just some information that is set when the driver registers the first time. The reason for that is that this information should be an init time data. If we keep asking for this information every time we are saying that this information can dynamically change which I assume we cannot handle in libnetwork. For example if the driver said it is global scoped while creating the network and said local scope when deleting the network it would cause confusion.

mavenugo replied:
Agreed. This calls for a change in controller data-structure to maintain the capabilities for each driver being registered. That results in larger change on all the places `network.driver` was used. Hence took this simpler approach. But, since you feel this can be a potential and real problem, I will make the change of maintaining this state in controller and keep the code churn minimal.

mrjana commented on datastore/datastore.go:
```diff
@@ -37,23 +50,35 @@ type KV interface {
+	return chain[len(rootChain):], nil
+}
+
 var errNewDatastore = errors.New("Error creating new Datastore")
```
error strings should start in lower case in all error strings in this file

mrjana commented on endpoint.go:
```diff
@@ -107,9 +110,72 @@ type endpoint struct {
+}
+
+func (ep *endpoint) MarshalJSON() ([]byte, error) {
+	epMap := make(map[string]interface{})
```
You need to hold the endpint lock during the whole of this function. It's fine to hold it across `json.Marshal` because we know that it's a purely data handling operation

mavenugo replied:
Agreed.

mrjana commented on endpoint.go:
```diff
@@ -107,9 +110,72 @@ type endpoint struct {
+}
+
+func (ep *endpoint) UnmarshalJSON(b []byte) (err error) {
+	var epMap map[string]interface{}
```
Need to hold endpoint lock here again.

mavenugo replied:
There is no harm in holding the lock in Unmarshal. But it is not necessary I think. The very nature of the Unmarshal operation, it is operating on a json string and converting it to a new `endpoint` operation and this endpoint object is still in transient create state and there is no chance of another goroutine to corrupt the data. So, it is optional to lock the data here. But if you prefer to see a lock operation for consistency and readability, I can very well do that.

mrjana commented on endpoint.go:
```diff
@@ -133,6 +199,37 @@ func (ep *endpoint) Network() string {
 }
 
+func (ep *endpoint) Key() []string {
+	return []string{datastore.EndpointKeyPrefix, string(ep.network.id), string(ep.id)}
```
Need to hold endpoint lock

mrjana commented on endpoint.go:
```diff
@@ -133,6 +199,37 @@ func (ep *endpoint) Network() string {
+}
+
+func (ep *endpoint) KeyPrefix() []string {
+	return []string{datastore.EndpointKeyPrefix, string(ep.network.id)}
```
endpoint lock

mrjana commented on endpoint.go:
```diff
@@ -133,6 +199,37 @@ func (ep *endpoint) Network() string {
+
+func (ep *endpoint) networkIDFromKey(key []string) (types.UUID, error) {
+	if key == nil || len(key) < 3 || key[0] != datastore.EndpointKeyPrefix {
+		return types.UUID(""), fmt.Errorf("Invalid endpoint key : %v", key)
```
error string lower case

mrjana commented on endpoint.go:
```diff
@@ -133,6 +199,37 @@ func (ep *endpoint) Network() string {
+}
+
+func (ep *endpoint) Index() uint64 {
+	return ep.dbIndex
```
endpoint lock

mrjana commented on endpoint.go:
```diff
@@ -133,6 +199,37 @@ func (ep *endpoint) Network() string {
+}
+
+func (ep *endpoint) SetIndex(index uint64) {
+	ep.dbIndex = index
```
endpoint lock

mrjana commented on endpoint.go:
```diff
@@ -107,9 +110,72 @@ type endpoint struct {
+	ep.name = epMap["name"].(string)
+	ep.id = types.UUID(epMap["id"].(string))
+
+	ib, _ := json.Marshal(epMap["ep_iface"])
```
What's the trick here? You are marshalling something in an unmarshal function and then unmarshalling the result?

mavenugo replied:
Just a convenient way to get a []byte of any structured data that is treated as interface{} which is then used to feed into UnMarshal again for that specific type.

mrjana commented on endpoint.go:
```diff
@@ -133,6 +199,37 @@ func (ep *endpoint) Network() string {
+	if key == nil || len(key) < 3 || key[0] != datastore.EndpointKeyPrefix {
+		return types.UUID(""), fmt.Errorf("Invalid endpoint key : %v", key)
+	}
+	return types.UUID(key[1]), nil
```
We need to get some structure to the key positions here. What does `key[1]` mean? If not may be a comment on the key structure would go a long way

mavenugo replied:
Agreed. I will try to find a way out.

mrjana commented on endpoint.go:
```diff
@@ -235,11 +332,13 @@ func (ep *endpoint) Join(containerID string, options ...EndpointOption) error {
 			ep.container = nil
+			ep.Unlock()
+		} else {
+			ep.network.ctrlr.addEndpointToStore(ep)
```
I don't think adding to data store operation should be handled in defer function. For starters we are not checking if there was an error while adding to store and if there was what are we going to do about that.

mrjana replied:
Also I am pretty sure you need to update the join status to endpoint store as an atomic update immediately after you construct the local container info. Otherwise there is a race if two containers in different hosts try to join. If addEndpointToStore fails atomic update(because of index being old) we need to re-update endpoint data from kv store which might reveal that a join already happened somewhere else and handle that case here accordingly.

mrjana commented on endpoint.go:
```diff
@@ -328,6 +427,7 @@ func (ep *endpoint) Leave(containerID string, options ...EndpointOption) error {
 	}
 	ep.container = nil
 	ep.Unlock()
+	ep.network.ctrlr.addEndpointToStore(ep)
```
This should be called after we get the `ctrlr` pointer while holding the network lock and then we invoke `addEndpointToStore`

mrjana commented on endpoint.go:
```diff
@@ -342,17 +442,29 @@ func (ep *endpoint) Leave(containerID string, options ...EndpointOption) error {
+		return err
 	}
 
+	if err := ep.network.ctrlr.deleteEndpointFromStore(ep); err != nil {
```
Should retrieve network and controller holding the respective locks and then `deleteEndpointFromStore` can be called

mrjana replied:
Also need to try to deleteEndpointFromStore atomically before triggering local driver delete because our local cache may be stale.

mrjana commented on network.go:
```diff
@@ -205,15 +246,14 @@ func (n *network) CreateEndpoint(name string, options ...EndpointOption) (Endpoi
+		return nil, err
+	}
+
+	if err := n.ctrlr.addEndpointToStore(ep); err != nil {
```
Again we probably need to rollback if this fails

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+}
+
+func (c *controller) newNetworkFromStore(n *network) error {
+	c.Lock()
```
Wrong long. Should hold network lock

mrjana commented on network.go:
```diff
@@ -176,7 +180,22 @@ func (n *network) Delete() error {
+		return err
+	}
+
+	if err = n.ctrlr.deleteNetworkFromStore(n); err != nil {
```
Again `deleteNetworkFromStore` needs to be done atomically from data store before local driver delete is triggered and if the atomic `deleteNetworkFromStore` fails because of stale index an immediate re-update of network endpoint data needs to happen. Which brings the question since networks and endpoints are handled as two separate tables how is this handled in terms of atomically determining how many endpoints are bound to this network?

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+		return nil
+	}
+
+	if err := cs.DeleteObject(n); err != nil {
```
This should be `DeleteObjectAtomic` and if it fails because of stale index it needs to be handled by a re-update.

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+		return err
+	}
+
+	if err := cs.DeleteTree(&endpoint{network: n}); err != nil {
```
This whole thing probably needs a rework. We cannot just like that delete the whole endpoint tree. In fact we should not delete the network if there are any active endpoints anywhere else

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+		return nil
+	}
+
+	if err := cs.DeleteObject(ep); err != nil {
```
Again this needs to be `DeleteObjectAtomic` and in case of failure needs to be properly handled.

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+	cs := c.store
+	c.Unlock()
+
+	nwPairs, err := cs.KVStore().WatchTree(datastore.Key(datastore.NetworkKeyPrefix), c.stopChan)
```
You don't seem to be using the `stopChan`. So why pass it?

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+			select {
+			case nws := <-nwPairs:
+				for _, kve := range nws {
+					var n network
```
When you get a notification is it always an add? How do you get delete notifications? Actually I don't see any code which pertains to how delete and delete notifications are handled?

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+				}
+			case eps := <-epPairs:
+				for _, epe := range eps {
+					var ep endpoint
```
Same here. How are delete endpoint notifications handled?

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+					c.Unlock()
+					if ok {
+						// Skip existing network update
+						if existing.dbIndex != n.dbIndex {
```
You need to hold `existing`'s lock while updating `dbIndex`

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+		return false
+	}
+	c.Lock()
+	n, ok := c.networks[ep.network.id]
```
Need to refactor this code because you need to hold network lock to access `id`

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+	}
+
+	ee := existing.(*endpoint)
+	if ee.dbIndex != ep.dbIndex {
```
Need to hold `ee` Lock while checking and updating `dbIndex`

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+	return ep.networkIDFromKey(eKey)
+}
+
+func (c *controller) skipEndpointUpdate(ep *endpoint) bool {
```
This function seems to be overloading two things. Checking if we have to skip endpoint update and updating the join state although the function name is misleading. May be need to change the name to something more appropriate.

mrjana commented on store.go:
```diff
@@ -0,0 +1,273 @@
+	return err
+}
+
+func (c *controller) addEndpointToStore(ep *endpoint) error {
```
I think this function should be named `updateEndpointToStore` since it is used for both add and update

mavenugo commented:
@mrjana @aboch Thanks a lot for your thorough review. Have addressed most of the comments, except for the Atomicity issue that Jana spotted with regards to Delete and Join operations. Being a big PR, I would request you to review the changes and make sure your other comments are addressed appropriately. I will work on the atomicity issue in parallel.

aboch commented on datastore/datastore.go:
```diff
@@ -37,23 +50,37 @@ type KV interface {
+
+var errNewDatastore = errors.New("error creating new datastore")
+var errInvalidConfiguration = errors.New("invalid configuration passed to datastore")
+var errInvalidAtomicRequest = errors.New("invalid atomic request")
```
Given the datastore component will be used heavily across the code, I think we should provide some context in the datastore generated error descriptions.

aboch commented on datastore/datastore.go:
```diff
@@ -116,5 +141,35 @@ func (ds *datastore) putObjectWithKey(kvObject KV, key ...string) error {
+// DeleteObjectAtomic performs atomic delete on a record
+func (ds *datastore) DeleteObjectAtomic(kvObject KV) error {
+	if kvObject == nil {
+		return errors.New("kvObject is nil")
```
Like here, this error description needs some context

aboch commented on datastore/datastore.go:
```diff
@@ -116,5 +141,35 @@ func (ds *datastore) putObjectWithKey(kvObject KV, key ...string) error {
 	if kvObjValue == nil {
 		return errors.New("Object must provide marshalled data for key : " + Key(kvObject.Key()...))
```
Let's adhere to start the error messages with lower case. Also please add some context.

mavenugo commented:
@mrjana @aboch addressed all the open comments. PTAL.

aboch commented:
LGTM

mrjana commented:
LGTM
